### ISSUE TITLE
RPM macros

### ISSUE THREAD

**Issue by korewaChino** [rpm-rs/rpm#59]
Could we add RPM macro support for rpm-rs? This will make packaging somewhat easier as we can just evaluate them on build time.

We can also use it for dynamic install paths, like `_libdir` or `_bindir`

**Comment by korewaChino** [comment]
I'm currently working on this with @windowsboy111 under the [andaman](https://github.com/FyraLabs/anda/tree/w/rpmspec) project.

We are currently in the process of re-implementing the RPM Lua interpreter using `rlua`.

**Comment by windowsboy111** [comment]
I've just finished most of an RPM spec parser. Take a look here: https://github.com/FyraLabs/anda/blob/w/rpmspec/rpmspec-rs/src/parse.rs

The only things I haven't done (afaik):
- `%if`, `%ifarch` and similar ones
- some lua functions
- some builtin functions

**Comment by windowsboy111** [comment]
`rpmspec-rs` currently simply does one job: it parses rpm spec files (only), so… we will need to implement more stuff and write more code to actually build an RPM from scratch.

Alternatively we can hand over the process to `rpmbuild`, but by this point we might as well make everything in rust.

**Comment by korewaChino** [comment]
> I'm not 100% clear on how deep the integration needs to be between the building process and the specfile parsing. Is the existing API or some reasonable extension of it sufficient to fully build an RPM from a specfile, or is there a need for significantly more extensive customization of the build process?

There is no building process as of yet. This library only parses RPM spec files, but however one can also create a tool combining both of this to create a full replacement for `rpmbuild`.

**Comment by dralley** [comment]
>rpmspec-rs currently simply does one job: it parses rpm spec files (only), so… we will need to implement more stuff and write more code to actually build an RPM from scratch.

> Alternatively we can hand over the process to rpmbuild, but by this point we might as well make everything in rust.

I was just assuming that if you went as far as to implement a specfile parser, that you were planning on implementing something similar to `rpmbuild` with it.  As far as I know, `rpmlib` simply takes a specfile as input, it doesn't really have an external API for constructing RPMs without a specfile like we do.

If there's a worthwhile purpose for having specfile parsing separate from building an RPM from a specfile, then we could have two additional repos.  But if the former is just something you need to implement the latter, then we'd just have one extra repo for an `rpmbuild` equivalent.

Alternatively we could shove all of this into one workspace.  But I would prefer individual repos, I think.

**Comment by windowsboy111** [comment]
I reckon we should create a new repo for rpmspec and another one for building rpms then.

**Comment by dralley** [comment]
Just curious, what is the use case you see for an independent specfile parsing library?

**Comment by windowsboy111** [comment]
Maybe something like `rpmlint` but in rust?

**Comment by dralley** [comment]
Sure, makes sense.  I don't necessarily think that trying to replace `rpmlint` would be worthwhile, but the fact that there's a potential usecase is enough.

**Comment by windowsboy111** [comment]
I'm currently just tidying up (the `%if` parsing part has been done this morning), please notify us once the new repo is ready.

**Comment by dralley** [comment]
@drahnr will need to do it, I don't have the necessary permissions

**Comment by drahnr** [comment]
@dralley @windowsboy111 I'll create a repo shortly, and give @dralley permissions to do so in the future

**Comment by windowsboy111** [comment]
`rpmspec-rs` since it parses spec files.

**Comment by drahnr** [comment]
Closing, if there are particular topics left to discuss, please open a discussion thread or a new issue :)

Thank you!

**Comment by windowsboy111** [comment]
It seems to me that I don't have write permissions…?
![image](https://github.com/rpm-rs/rpm/assets/48051237/308be54a-badf-47ab-b79e-12a979e398ae)
I tried `git push` but it got denied.

**Comment by windowsboy111** [comment]
Nevermind, had to accept the invite. My bad!
